### PR TITLE
Switch to Matrixing for Faster Build Checks and Full Config Validation

### PR DESCRIPTION
**Pull request by barndawgie**

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -7,36 +7,68 @@ on:
     branches:
       - main
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.ref_name }}
```
Comment (tronikos) [inline .github/workflows/build.yml:11]: Does this need github.event.pull_request.number?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -7,36 +7,68 @@ on:
...
+            ./${{ matrix.config_file }}
+      - name: Validate configuration
+        run: esphome -s external_components_source components config ./${{ matrix.config_file }}
+
+  build:
+    name: Build ${{ matrix.appliance.name }} Firmware
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        appliance:
```
Comment (tronikos) [inline .github/workflows/build.yml:57]: Similar to the other PR, is it possible to define this map once in this file? Is it possible to define it once for all workflows?

File: .github/workflows/build.yml (hunk 1 above)
Comment (barndawgie) [reply .github/workflows/build.yml:11]: Probably right.

File: .github/workflows/build.yml (hunk 2 above)
Comment (barndawgie) [reply .github/workflows/build.yml:57]: I can't find a way - we can try to refactor in a future update.

**Review by tronikos — APPROVED** [review]

**Pull request merged**